pkg/security/probe/sysctl/snapshot.go

```diff
@@ -0,0 +1,166 @@
+		"stable_secret",
+	}
+	redactedContent = "******"
+)
+
+// readFileContent reads a file and processes its content based on the given rules.
+func readFileContent(file string) (string, error) {
+	if slices.Contains(ignoredBaseNames, path.Base(file)) {
+		return redactedContent, nil
+	}
+
+	data, err := os.ReadFile(file)
+	if err != nil {
+		return "", err
+	}
+
+	return string(data), nil
+}
+
+// SnapshotEvent is a wrapper used for serialization
+type SnapshotEvent struct {
+	Sysctl Snapshot `json:"sysctl"`
+}
+
+// NewSnapshotEvent returns a new sysctl snapshot event
+func NewSnapshotEvent() (*SnapshotEvent, error) {
+	se := &SnapshotEvent{
+		Sysctl: NewSnapshot(),
+	}
+	if err := se.Sysctl.Snapshot(); err != nil {
+		return nil, err
+	}
+	return se, nil
+}
+
+// ToJSON serializes the current SnapshotEvent object to JSON
+func (s *SnapshotEvent) ToJSON() ([]byte, error) {
+	return json.Marshal(s)
+}
+
+// Snapshot defines an internal core dump
+type Snapshot struct {
+	// Proc contains the /proc system control parameters and their values
+	Proc map[string]interface{} `json:"proc,omitempty"`
+	// Sys contains the /sys system control parameters and their values
+	Sys map[string]interface{} `json:"sys,omitempty"`
+}
+
+// NewSnapshot returns a new sysctl snapshot
+func NewSnapshot() Snapshot {
+	return Snapshot{
+		Proc: make(map[string]interface{}),
+		Sys:  make(map[string]interface{}),
+	}
+}
+
+// Snapshot runs the snapshot by going through the filesystem
+func (s *Snapshot) Snapshot() error {
+	if err := s.snapshotProcSys(); err != nil {
+		return fmt.Errorf("couldn't snapshot /proc/sys: %w", err)
+	}
+
+	if err := s.snapshotSys(); err != nil {
+		return fmt.Errorf("coudln't snapshot /sys: %w", err)
+	}
+	return nil
+}
+
+// snapshotProcSys recursively reads files in /proc/sys and builds a nested JSON structure.
+func (s *Snapshot) snapshotProcSys() error {
+	return filepath.Walk("/proc/sys", func(file string, info fs.FileInfo, err error) error {
+		if err != nil {
+			return err
+		}
+
+		// Skip directories
+		if info.IsDir() {
+			return nil
+		}
+
+		// Skip if mode doesn't allow reading
+		mode := info.Mode()
+		if mode&0400 == 0 && mode&0040 == 0 && mode&0004 == 0 {
+			return nil
+		}
+
+		relPath, err := filepath.Rel("/proc", file)
+		if err != nil {
+			return err
+		}
+
+		value, err := readFileContent(file)
+		if err != nil {
+			return nil // Skip files that can't be read
+		}
+
+		s.InsertSnapshotEntry(s.Proc, relPath, value)
+		return nil
+	})
+}
+
+// snapshotSys reads security relevant files from the /sys filesystem
+func (s *Snapshot) snapshotSys() error {
+	lockdownValue, err := readFileContent("/sys/kernel/security/lockdown")
+	if err != nil {
+		return err
+	}
+	s.InsertSnapshotEntry(s.Sys, "kernel/security/lockdown", lockdownValue)
+
+	lsmValue, err := readFileContent("/sys/kernel/security/lsm")
+	if err != nil {
+		return err
+	}
+	s.InsertSnapshotEntry(s.Sys, "kernel/security/lsm", lsmValue)
+	return nil
+}
```

Comment:
Shouldn't we make use of `SysFSRoot()` for these paths?

Reply:
Also, shouldn't we ignore the error in case a file doesn't exist? Otherwise this will cause the `NewSnapshotEvent()` function to fail

Reply:
Yup and yup, on it 👍🏻

Reply:
Actually thinking about it: I'd rather not send any `sysctl_snapshot` event than sending a half backed one (because you won't really know what you can rely on). Happy to discuss if you disagree ! :)